File: src/partition/build_rcells.f90
! Copyright (C) 2008-2010 by Yung-Yu Chen.  See LICENSE.txt for terms of usage.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
! subroutine build_rcells:
!     Build unstructured grid cell relation accroding to face-cell 
!     relation. 
!     Return rcells and rcellno.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
subroutine build_rcells(max_clnfc, nface, ncell, clfcs, fccls, rcells, rcellno)
implicit none
! arguments.
integer(4), intent(in) :: max_clnfc, nface, ncell
integer(4), intent(in), dimension(0:max_clnfc, 0:ncell-1) :: clfcs
integer(4), intent(in), dimension(0:3, 0:nface-1) :: fccls
integer(4), intent(out), dimension(0:max_clnfc-1, 0:ncell-1) :: rcells
integer(4), intent(out), dimension(0:ncell-1) :: rcellno
! iterators.
integer(4) :: icl, it, ifc
rcells  = -1
rcellno = 0
do icl = 0, ncell-1
    do it = 0, clfcs(0,icl)-1
        ifc = clfcs(it+1,icl)
        ! is not a face; shouldn't happen
        if( ifc .eq. -1 ) then
            rcells(it,icl) = -1
            cycle
        else if( fccls(0,ifc) .eq. icl ) then
            ! has neibor block
            if( fccls(2,ifc) .ne. -1 ) then
                rcells(it,icl) = -1
            ! is interior
            else
                rcells(it,icl) = fccls(1,ifc)
            end if
        ! I am the neiboring cell
        else if( fccls(1,ifc) .eq. icl ) then
            rcells(it,icl) = fccls(0,ifc)
        end if
        ! count rcell number
        if( rcells(it,icl) .ge. 0 ) then 
            rcellno(icl) = rcellno(icl) + 1
        else
            rcells(it,icl) = -1
        end if
    end do
end do
end subroutine build_rcells
! vim:set nu et tw=72 ts=4 sw=4 cino=>4:

File: src/partitionc/build_rcells.c
// Copyright (C) 2008-2010 by Yung-Yu Chen.  See LICENSE.txt for terms of
// usage.
#include "solvcon.h"
int build_rcells(MeshData *msd, int *rcells, int *rcellno) {
    // pointers.
    int *pclfcs, *pfccls;
    int *prcells;
    // iterators.
    int icl, ifl, ifl1, ifc;

    // initialize.
    prcells = rcells;
    for (icl=0; icl<msd->ncell; icl++) {
        for (ifl=0; ifl<CLMFC; ifl++) {
            prcells[ifl] = -1;
        };
        rcellno[icl] = 0;
        // advance pointers.
        prcells += CLMFC;
    };
    
    // count.
    pclfcs = msd->clfcs;
    prcells = rcells;
    for (icl=0; icl<msd->ncell; icl++) {
        for (ifl=1; ifl<=pclfcs[0]; ifl++) {
            ifl1 = ifl-1;
            ifc = pclfcs[ifl];
            pfccls = msd->fccls + ifc*FCREL;
            if (ifc == -1) {    // NOT A FACE!? SHOULDN'T HAPPEN.
                prcells[ifl1] = -1;
                continue;
            } else if (pfccls[0] == icl) {
                if (pfccls[2] != -1) {  // has neighboring block.
                    prcells[ifl1] = -1;
                } else {    // is interior.
                    prcells[ifl1] = pfccls[1];
                };
            } else if (pfccls[1] == icl) {  // I am the neighboring cell.
                prcells[ifl1] = pfccls[0];
            };
            // count rcell number.
            if (prcells[ifl1] >= 0) {
                rcellno[icl] += 1;
            } else {
                prcells[ifl1] = -1;
            };
        };
        // advance pointers.
        pclfcs += CLMFC+1;
        prcells += CLMFC;
    };

    return 0;
};
// vim: set ts=4 et:
